BUG: Fixed slicing of chararrays on Python 3.

embray opened the pull request:
When taking a slice of a chararray it was calling the rstrip() method
on the resulting slice, resulting in a new array rather than a view
of the original.  This was an unintended consequence of the `sq_slice`
member of the `tp_as_sequence` mapping being ignored in Python 3, so
that slice lookups go directly through `__getitem__`.

I realize the chararray type is deprecated now which is fine, but this bug
introduced a regression for the `astropy.io.fits` module on Python 3 (I'm not
sure why this only seems to have broken recently, but it did).  So it can't hurt
to fix this, and in the meantime I'm going to work on eliminating the dependency
on chararray in Astropy.

embray committed:
BUG: Fixed slicing of chararrays on Python 3.

When taking a slice of a chararray it was calling the rstrip() method
on the resulting slice, resulting in a new array rather than a view
of the original.  This was an unintended consequence of the sq_slice
member of the tp_as_sequence mapping being ignored in Python 3, so
that slice lookups go directly through __getitem__.

Fix test_expandtabs to not make the assumption that rstrip() will be applied twice to a value when using T[x][y] style indexing.

charris commented:
Hmm, universal test failures.

embray commented:
Not sure what happened on that last commit. I  thought I had tested it.  I'll come back to it tomorrow.

mhvk commented on numpy/core/tests/test_defchararray.py:
```diff
@@ -629,6 +629,22 @@ def test_rmod(self):
+        assert sl2.base is arr
+        assert sl2.base.base is arr.base
+
+        assert arr[0, 0] == b'abc'
```
This test fails on python3, because the array contains unicode; add `, dtype='S'` in the definition.

embray replied:
Ah, righto.

mhvk commented:
The other test that fails is in  https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_defchararray.py#L332
```
 assert_(T[2][0] == asbytes('123      345'))
```
Here, the problem is that that element contains:
```
b'123 \t 345 \x00 '
```
Previously, taking the first index `T[2]` stripped both the last space and the `0` as `defchararray.rstrip` was called, which creates a new array in which the `\0` disappears. But now, one calls `bytes.rstrip`, which does not remove the `\0`. 

I think the old behaviour really tested the buggy behaviour, so would suggest simply adapting the test.

mhvk commented:
p.s. Your first commit happened not to hit that bug since `T[2]` was incorrectly treated as if a single element was taken (indeed, that one definitely tests the wrong thing!).

embray commented:
@mhvk Thanks for looking into that.  Had I known you were doing so I wouldn't have wasted the last 15 minutes doing the same :)  I'll have a fix shortly (I agree with you that the original `test_expandtabs`, as defined, was expecting behavior w.r.t. rstrip() that I think is faulty, so I just fixed the test).

embray commented:
As someone else raised elsewhere (I forget exactly) I updated the fix to be the same as the workaround I've been using for a long time in PyFITS.  I also updated the tests as pointed out by @mhvk.

If / when this is decided ready to merge I'll squash the changes as well.

mhvk commented:
@embray - sorry for the duplcated work!  With numpy 1.10 getting close, I'm trying to ensure that we can use astropy as a nice test bed...

charris commented:
Yes, 1.10 is Numpy for the stars! Note that chararray was maintained by someone at STSci and with FITS now the responsibility of AstroPy, that reponsibility probably devolves to you.

charris commented:
@embray Can be squashed anytime now.

embray commented:
@mhvk No need to apologize--it's encouraging that we both came to the same conclusion.

@charris Had no idea chararray came from us, though I'm not too surprised given that it's a numarray legacy.  I'm happy to help kill it as soon as possible.

charris commented:
Thanks Eric.
